File: yaml_parser.py
def parse_value(value):
    """
    Parses a string value and converts it to the appropriate type.
    
    Args:
        value (str): The string value to parse
    
    Returns:
        The value converted to the appropriate type (int, float, bool, or str)
    """
    value = value.strip()

    if value.lower() == 'true':
        return True
    elif value.lower() == 'false':
        return False

    elif value.lower() == 'null':
        return None

    try:
        if '.' in value:
            return float(value)
        else:
            return int(value)
    except ValueError:
        pass
    
    # Return the value as a string if no other type matches
    return value

def parse_decorated_pair(current_group, metadata_comment, key_value_pair):
    """
    Parses a decorated key-value pair with its metadata comment into a structured dictionary.
    
    Args:
        current_group (str): The current group name
        metadata_comment (str): Metadata comment starting with #$ 
        key_value_pair (str): YAML key-value pair
    
    Returns:
        dict: Parsed dictionary with group, id, value, and metadata attributes
    """
    parsed_pair = {}
    
    # Split attributes, skipping the first element (the #$ marker)
    attributes = metadata_comment.split()[1:]
    
    # Parse key and value
    key, value = key_value_pair.split(': ', 1)
    
    # Basic parsing
    parsed_pair['group'] = current_group
    parsed_pair['id'] = key.strip()
    parsed_pair['value'] = parse_value(value)
    parsed_pair['comment'] = metadata_comment
    
    # Parse additional attributes
    for attr in attributes:
        if ':' in attr:
            attr_key, attr_value = attr.split(':', 1)
            parsed_pair[attr_key] = parse_value(attr_value)
    
    return parsed_pair

def parse_yaml(file_contents):
    """
    Parses a configuration file with decorated YAML formatted data.
    
    Args:
        file_contents (str): Contents of the YAML file
    
    Returns:
        dict: Dictionary with 'decorated' and 'plain' YAML data
    """
    # Split file contents into lines
    lines = file_contents.split('\n')
    
    # Data containers
    decorated_data = {}
    plain_data = {}
    
    # Track current group
    current_group = None
    
    # Iterate through lines
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        
        # Reset current group on empty lines
        if not line:
            current_group = None
            i += 1
            continue
        
        # Handle group declarations
        if line.startswith('#$ group'):
            if i + 1 < len(lines):
                current_group = lines[i + 1].strip()[:-1]  # Remove trailing colon
                decorated_data[current_group] = []
                i += 2
                continue
        
        # Handle decorated key-value pairs
        elif line.startswith('#$'):
            if i + 1 < len(lines):
                metadata_comment = line
                key_value_pair = lines[i + 1].strip()
                
                if current_group:
                    decorated_pair = parse_decorated_pair(current_group, metadata_comment, key_value_pair)
                    decorated_data[current_group].append(decorated_pair)
                i += 2
                continue
        
                # Skip comments
        elif line.startswith('#'):
            i += 1
            continue
        
        # Add non-comment lines to plain data
        current_group = line.strip()[:-1]  # Remove trailing colon
        plain_data[current_group] = {}
        
        i += 1
        while i < len(lines) and lines[i]:
            key_value_pair = lines[i].strip()
            key, value = key_value_pair.split(':', 1)
            plain_data[current_group][key.strip()] = parse_value(value.strip())
            i += 1
            
        i += 1
    
    return {
        'decorated': decorated_data,
        'plain': plain_data
    }

def rebuild_yaml(parsed_config):
    """
    Rebuilds a configuration file from parsed data.
    
    Args:
        parsed_config (dict): Parsed configuration dictionary
    
    Returns:
        str: Rebuilt configuration file content
    """
    rebuilt_config = ''
    
    # Rebuild decorated groups
    for group, pairs in parsed_config['decorated'].items():
        rebuilt_config += '#$ group\n'
        rebuilt_config += f'{group}:\n'
        
        for pair in pairs:
            rebuilt_config += f'  {pair["comment"]}\n'
            rebuilt_config += f'  {pair["id"]}: {pair["value"]}\n'
        
        rebuilt_config += '\n'  # Empty line between groups
        
    # Append plain YAML content
    for group, pairs in parsed_config['plain'].items():
        rebuilt_config += f'{group}:\n'
        for key, value in pairs.items():
            rebuilt_config += f'  {key}: {value}\n'
            
        rebuilt_config += '\n'  # Empty line between groups
    
    rebuilt_config = rebuilt_config[:-2] # hack to remove trailing blank lines

    # Return as string
    return rebuilt_config

File: server/scripts/yaml-parser.js
// const fs = require('fs');

/**
 * Parses a decorated key-value pair with its metadata comment into a structured object.
 * @param {string} metadataComment - The metadata comment starting with #$ (e.g., "#$ tag:input type:number")
 * @param {string} keyValuePair - The YAML key-value pair following the metadata (e.g., "frequency: 50")
 * @returns {Object} Parsed object containing id, value, and all metadata attributes
 */
function parseDecoratedPair(currentGroup, metadataComment, keyValuePair) {
    const parsedPair = {};
    const attributes = metadataComment.split(/\s+/).slice(1);

    parsedPair.group = currentGroup;
    parsedPair.id = keyValuePair.split(': ')[0];
    parsedPair.value = keyValuePair.split(': ')[1];
    parsedPair.comment = metadataComment;  // Store original comment for rebuilding

    // Parse each attribute (tag:input, type:number, etc.) into key-value pairs
    attributes.forEach(e => { parsedPair[e.split(':')[0]] = e.split(':')[1]; });

    return parsedPair;
}

/**
 * Parses a configuration file containing decorated YAML formatted data.
 * Supports grouped elements with metadata comments describing how an element should be displayed, and plain 
 * YAML content.
 * @param {string} file - Path to the configuration file
 * @returns {Object} Object containing decorated (grouped) and plain YAML data
 */
export function parseYaml(fileContents) {
    const lines = fileContents.split('\n');
    const decoratedData = {};  // Stores grouped decorated elements
    const plainData = [];      // Stores regular YAML lines

    // Track the current group nest level we are in
    let currentGroup = null;

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];

        // Reset current group on empty lines (group separator)
        if (!line) { currentGroup = null; continue; }

        // Handle decorated lines (starting with #$)
        if (line.trim().startsWith('#$')) {
            // Handle group declarations
            if (line === '#$ group') {
                const nextLine = lines[i + 1];
                currentGroup = nextLine.trim().slice(0, -1);  // Remove trailing colon
                decoratedData[currentGroup] = [];
                i++;  // Group name already used, so skip in next iteration
                continue;
            } else {
                // Handle decorated key-value pairs
                const metadataComment = line.trim();
                const keyValuePair = lines[i + 1]?.trim();
                if (currentGroup) {
                    decoratedData[currentGroup].push(parseDecoratedPair(currentGroup, metadataComment, keyValuePair));
                } else {
                    console.error('Error: No group defined for decorated key-value pair: "' + keyValuePair + '"');
                }
                i++;  // Next key-value already used, so skip in next iteration
                continue;
            }
        }

        // TODO: Handle regular comments. Skip for now.
        if (line.trim().startsWith('#')) continue;

        // Add non-comment lines to plainData
        plainData.push(line);
    }
    
    return { decorated: decoratedData, plain: plainData };
}

/**
 * Rebuilds a configuration file from parsed data.
 * @param {Object} parsedConfig - The parsed configuration object containing decorated and plain data
 * @returns {string} Rebuilt configuration file content
 */
export function rebuildYaml(parsedConfig, file) {
    let rebuiltConfig = '';

    // Rebuild decorated groups
    for (const group in parsedConfig.decorated) {
        rebuiltConfig += '#$ group\n' + group + ':\n';
        parsedConfig.decorated[group].forEach(pair => {
            rebuiltConfig += '  ' + pair.comment + '\n' + '  ' + pair.id + ': ' + pair.value + '\n';
        });
        rebuiltConfig += '\n';
    }

    // Append plain YAML content
    rebuiltConfig += parsedConfig.plain.join('\n');

    return rebuiltConfig;
}

// Example usage
// const file = 'config.yaml';
// const fileContents = fs.readFileSync(file, 'utf8');
// parsedConfig = parseConfig(fileContents);
// console.log(parsedConfig);

// rebuiltConfig = rebuildConfig(parsedConfig, file);
// fs.writeFileSync(file, rebuiltConfig);
// console.log(rebuiltConfig);

/**
Parsed Config Example:
    {
    decorated: {
        transmitter: [
            {
            id: 'frequency',
            value: '50',
            comment: '#$ tag:input type:range min:1 max:100 step:1',
            tag: 'input',
            type: 'range',
            min: '1',
            max: '100',
            step: '1'
            },
        ],
        receiver: [
            {
            id: 'power',
            value: '5',
            comment: '#$ tag:input type:number min:0 max:10 step:1',
            tag: 'input',
            type: 'number',
            min: '0',
            max: '10',
            step: '1'
            }
        ]
    },
    plain: [ 'pinset:', '  pin1: 1', '  pin2: 2' ]
    }
 **/